File: remove-linked-list-elements/remove-linked-list-elements.cc
#include <bits/stdc++.h>
using namespace std;

struct ListNode {
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(NULL) {}
};
typedef ListNode *node;
class Solution {
public:
  node removeElements(node head, int val) {
    ListNode dummy(0);
    node prev = &dummy;
    dummy.next = head;
    while(head) {
      node next = head->next;
      if(head->val == val) {
        prev->next = head->next;
      } 
      else prev = head;
      head = next;
    }
    return dummy.next;
  }
};

int main(int argc, const char *argv[])
{
  Solution so;
  ListNode n0(0), n1(1);
  node h = &n0;
  n0.next = &n1;
  assert(so.removeElements(h, 1) == &n0 && n0.next == NULL);
  return 0;
}

File: rs/src/quests/remove_linked_list_elements.rs
/**
 * [203] Remove Linked List Elements
 *
 * Given the head of a linked list and an integer val, remove all the nodes of the linked list that has Node.val == val, and return the new head.
 
Example 1:
Input: head = [1,2,6,3,4,5,6], val = 6
Output: [1,2,3,4,5]

Example 2:
Input: head = [], val = 1
Output: []

Example 3:
Input: head = [7,7,7,7], val = 7
Output: []

 
Constraints:

	The number of nodes in the list is in the range [0, 104].
	1 <= Node.val <= 50
	0 <= val <= 50

 */
pub struct Solution {}
use super::super::util::linked_list::{ListNode, to_list};

// submission codes start here

// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn remove_elements(mut head: Option<Box<ListNode>>, val: i32) -> Option<Box<ListNode>> {
        let mut dummy = None;
        let mut tail = &mut dummy;
        while let Some(mut node) = head.take() {
            head = node.next.take();
            if node.val != val {
                tail = &mut tail.insert(node).next;
            }
        }
        dummy
    }
}

// submission codes end

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() {
    }
}
